proposals/csharp-7.0/tuples.md

```diff
@@ -0,0 +1,437 @@
+
+```csharp
+var array = new (int x, int y)[10];
+var nullable = new (int x, int y)?();
+```
+
+Tuples can be created for a known target type:
+
+```csharp
+public (int sum, int count) Tally(IEnumerable<int> values)
+{
+    var s = 0; var c = 0;
+    foreach (var value in values) { s += value; c++; }
+    return (s, c); // target typed to (int sum, int count)
+}
+```
+
+Specifying field names is optional. Duplicate names are disallowed.
+
+```csharp
+var t1 = (sum: 0, count: 1); // OK, all fields are named
+var t2 = (sum: 0, 1);       // Ok, implies (int sum, int);
+var t3 = (sum: 0, sum: 1);  // error! duplicate names.
+```
+
+### Tuple literals
+
+```csharp
+var t1 = (0, 2);              // infer tuple type from values
+var t2 = (sum: 0, count: 1);  // infer tuple type from names and values
+```
+
+A tuple literal is "target typed" whenever possible. The tuple literal has a "conversion from expression" to any tuple type, as long as the element expressions of the tuple literal have an implicit conversion to the element types of the tuple type.
+
+```csharp
+(string name, byte age) t = (null, 5); // Ok: the expressions null and 5 convert to string and byte
+```
+
+In cases where the tuple literal is not part of a conversion, it acquires its "natural type", which means a tuple type where the element types are the types of the constituent expressions. Since not all expressions have types, not all tuple literals have a natural type either:
+
+```csharp
+var t = ("John", 5); // Ok: the type of t is (string, int)
+var t = (null, 5); //   Error: null doesn't have a type
+```
+
+A tuple literal may include names, in which case they become part of the natural type:
+
+```csharp
+var t = (name: "John", age: 5); // The type of t is (string name, int age)
+```
+
+Tuple literals may be deconstructed directly:
+
+```csharp
+(string x, byte y, var z) = (null, 1, 2);
+(string x, byte y) t = (null, 1);
+```
+
+Or for deconstructing assignment:
+
+```csharp
+string x;
+byte y;
+
+(x, y) = (null, 1);
+(x, y) = (y, x); // swap!
+```
+
+The evaluation order of deconstruction assignment expressions is "breadth first":
+
+1. Evaluate the LHS: Evaluate each of the expressions inside of it one by one, left to right, to yield side effects and establish a storage location for each.
+1. Evaluate the RHS: Evaluate each of the expressions inside of it one by one, left to right to yield side effects
+1. Convert each of the RHS expressions to the LHS types expected, one by one, left to right.
+1. Assign each of the conversion results from 3 to the storage locations found in 
+
+> **Note to reviewers**: I found this in the LDM notes for July 13-16, 2016. I don't think it is still accurate:
+
+A deconstructing assignment is a *statement-expression* whose type could be `void`.
+
+### Duality with underlying type
+
+Tuples map to underlying types of particular names.
```

Comment:
>map to underlying types of particular names [](start = 7, length = 43)

This section does not specify the mapping.  For example, VT`8 is not itself a tuple type unless the rest field is substituted with a tuple type.

Reply:
logged in #3151